File: Human_written_code/Code_with_comments/Python/Q4.py
def maxProfit(arr):
    maxPro = 0          # Initialize maximum profit
    minPrice = float('inf')  # Set initial minimum price to infinity

    for i in range(len(arr)):
        minPrice = min(minPrice, arr[i])  # Update minimum price if a lower price is found
        maxPro = max(maxPro, arr[i] - minPrice)  # Calculate max profit if sold at current price

    return maxPro  # Return the maximum profit

arr = [7, 1, 5, 3, 6, 4]
maxPro = maxProfit(arr)
print("Max profit is:", maxPro)

File: Human_written_code/Code_with_comments/CPP/Q4.cpp
#include<bits/stdc++.h>
using namespace std;

// Function to calculate the maximum profit
int maxProfit(vector<int> &arr) {
    int maxPro = 0;
    int minPrice = INT_MAX; // Track the minimum price so far

    for (int i = 0; i < arr.size(); i++) {
        minPrice = min(minPrice, arr[i]); // Update the minimum price
        maxPro = max(maxPro, arr[i] - minPrice); // Calculate maximum profit
    }
    
    return maxPro;
}

int main() {
    vector<int> arr = {7, 1, 5, 3, 6, 4};
    int maxPro = maxProfit(arr);
    cout << "Max profit is: " << maxPro << endl;

    return 0;
}
